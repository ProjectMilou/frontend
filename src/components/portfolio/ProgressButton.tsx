// adapted from https://github.com/mui-org/material-ui/blob/f285808fbf8728b0cc0435e65a76dbc691be3643/docs/src/pages/components/progress/CircularIntegration.tsx

import {
  Button,
  CircularProgress,
  makeStyles,
  PropTypes,
} from '@material-ui/core';
import React, { MouseEventHandler } from 'react';

export type ProgressButtonProps = {
  onClick: MouseEventHandler;
  disabled?: boolean;
  loading?: boolean;
  color?: PropTypes.Color;
  className?: string;
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
  },
  wrapper: {
    margin: theme.spacing(1),
    position: 'relative',
  },
  buttonProgress: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    marginTop: '-12px',
    marginLeft: '-12px',
  },
}));

/**
 * A button with a circular progress indicator.
 *
 * @param onClick - Click event handler
 * @param disabled - Whether the button is disabled. Note: The button is always disabled if {@param loading} is set.
 * @param loading - Whether the loading indicator should be displayed
 * @param color - The color of the button
 * @param className - Button class name(s)
 * @param children - The content of the button
 */
const ProgressButton: React.FC<ProgressButtonProps> = ({
  onClick,
  disabled,
  loading,
  color,
  className,
  children,
}) => {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <div className={classes.wrapper}>
        <Button
          disabled={loading || disabled}
          onClick={onClick}
          color={color}
          className={className}
        >
          {children}
        </Button>
        {loading && (
          <CircularProgress size={24} className={classes.buttonProgress} />
        )}
      </div>
    </div>
  );
};

export default ProgressButton;
