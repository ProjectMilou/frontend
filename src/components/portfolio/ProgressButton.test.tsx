import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import { PropsWithChildren } from 'react';
import ProgressButton, { ProgressButtonProps } from './ProgressButton';

describe('ProgressButton', () => {
  const defaultProps: PropsWithChildren<ProgressButtonProps> = {
    onClick: jest.fn(),
    children: 'Button',
  };

  const renderComponent = (
    newProps?: Partial<PropsWithChildren<ProgressButtonProps>>
  ) => {
    const props = { ...defaultProps, ...newProps };
    return {
      ...render(<ProgressButton {...props} />),
      props,
    };
  };

  test('render the button and handle click', () => {
    const { container, props } = renderComponent();
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('button')!);
    expect(props.onClick).toHaveBeenCalled();
  });

  test('render the loading button', () => {
    const { container, props } = renderComponent({ loading: true });
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('button')!);
    expect(props.onClick).not.toHaveBeenCalled();
  });
});
