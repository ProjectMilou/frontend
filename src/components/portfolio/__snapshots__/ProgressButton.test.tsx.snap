// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressButton render the button and handle click 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="makeStyles-wrapper-2"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Button
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`ProgressButton render the loading button 1`] = `
<div>
  <div
    class="makeStyles-root-4"
  >
    <div
      class="makeStyles-wrapper-5"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Button
        </span>
      </button>
      <div
        class="MuiCircularProgress-root makeStyles-buttonProgress-6 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 24px; height: 24px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
